microsoft/testsuites/nvme/performance.py

```diff
@@ -0,0 +1,128 @@
+    TIME_OUT = 3000
+
+    @TestCaseMetadata(
+        description="""
+        This test case uses fio to test NVMe disk performance.
+        """,
+        priority=2,
+        timeout=TIME_OUT,
+        requirement=simple_requirement(
+            supported_features=[NvmeSettings(disk_count=8)],
+        ),
+    )
+    def perf_nvme(self, node: Node, environment: Environment) -> None:
+        nvme = node.features[Nvme]
+        nvme_namespaces = nvme.get_namespaces()
+        disk_count = len(nvme_namespaces)
+        assert_that(disk_count).described_as(
+            "At least 1 NVMe disk for fio testing."
+        ).is_greater_than(0)
+        filename = ":".join(nvme_namespaces)
+        echo = node.tools[Echo]
+        # This will have kernel avoid sending IPI to finish I/O on the issuing CPUs
+        # if they are not on the same NUMA node of completion CPU.
+        # This setting will give a better and more stable IOPS.
+        for nvme_namespace in nvme_namespaces:
+            # /dev/nvme0n1 => nvme0n1
+            disk_name = nvme_namespace.split("/")[-1]
+            echo.write_to_file(
+                "0",
+                node.get_pure_path(f"/sys/block/{disk_name}/queue/rq_affinity"),
+                sudo=True,
+            )
+        cpu = node.tools[Lscpu]
+        core_count = cpu.get_core_count()
+        start_qdepth = core_count
+        max_qdepth = start_qdepth * 256
+        numjob = core_count
+        numjobiterator = 0
+        fio_result_list: List[FIOResult] = []
+        fio = node.tools[Fio]
+        for mode in FIOMODES:
+            qdepth = start_qdepth
+            while qdepth <= max_qdepth:
+                iodepth = int(qdepth / numjob)
+                fio_result = fio.launch(
+                    name=f"iteration{numjobiterator}",
+                    filename=filename,
+                    mode=mode.name,
+                    iodepth=int(iodepth),
+                    numjob=numjob,
+                )
+                fio_result_list.append(fio_result)
+                qdepth = qdepth * 2
+                numjobiterator += 1
+        fio_messages: List[FIOResultsMessage] = fio.parse_test_results(fio_result_list)
+        self._handle_and_send_back_results(
+            core_count, disk_count, environment, fio_messages
+        )
+
+    def _handle_and_send_back_results(
+        self,
+        core_count: int,
+        disk_count: int,
+        environment: Environment,
+        fio_messages: List[FIOResultsMessage],
+    ) -> None:
+        information: Dict[str, str] = environment.get_information()
+        host_type = information.pop("platform", "")
+        host_by = information.pop("location", "")
+        host_os = information.pop("host_version", "")
+        guest_distro = information.pop("distro_version", "")
+        guest_size = information.pop("vmsize", "")
+        kernel_version = information.pop("kernel_version", "")
+        lis_version = information.pop("lis_version", "")
```

Comment:
1. Please refer to `fields_to_dict` to create a method `dict_to_fields`. So it can be easier in future.
2. Use `get` instead of `pop`. The pop will remove the value from the dict, it doesn't necessary.

Reply:
updated